### ISSUE TITLE
TransformLayer should participate in raster caching

### ISSUE THREAD

**Issue by dnfield** [flutter/flutter#91648]
The framework uses `RepaintBoundary`s to push `OffsetLayer`s to signify an area that should paint independently. This gets turned into a `TransformLayer` on the engine side, with an `SkMatrix` that is translate-only.

This may result in multiple pictures that are individually raster cached, but often times a repaint boundary indicates an entire area of multiple pictures that would benefit from raster caching. This is particularly true in scrolling use cases, where repaint boundaries are used to animate the offset of a layer.

It should be possible for `flutter::TransformLayer`s to check if they are a translation only matrix and raster cache using the same heuristics as other layers today. The theory is that this will help with scrolling scenarios. We'll want to monitor impact on memory usage.

/cc @goderbauer @flar

**Comment by dnfield** [comment]
How do we handle children changing for opacity and image shader layers?

**Comment by goderbauer** [comment]
For context, I thought we could do something like this also on Transform layer:

https://github.com/flutter/engine/blob/a5aa1d98f502d5289fd8cb96e713ecb13e670ce2/flow/layers/image_filter_layer.cc#L72

https://github.com/flutter/engine/blob/9e9522c1692a62934d586992eb779e13b61bf55b/flow/layers/opacity_layer.cc#L62

**Comment by flar** [comment]
How does this request achieve that? All of that is already done (modulo SVG entries bypass some of our logic to individually cache them that is being worked on under other issues)?

**Comment by flar** [comment]
This sounds like a win. One thing to look for, though, would be if it causes jank while scrolling as the entire scrolled container is invalidated when a new child is created. Caching the scrolled items has reduced jank for that situation because only the new child needs to be cached.

Another thing to consider is that in the cache list shown above we see the contents of the items being cached, but the border of the item is not. That effect suggests that the RepaintBoundary is too low in the sub-tree. The scrollable container should literally be a list of RepaintBoundaries, but it looks like it might be a list of decorations that each wrap a RepaintBoundary. Do we see an improvement if the entire item is cached rather than just its contents?

This is also driving us towards a situation where we may end up with extra nested cache items. I brought this up in the PR for caching nested Picture objects. The issue is that every layer and picture is making an independent decision about whether it should be cached - which is fine if all decisions are on leaves. But, as noted last year in https://github.com/flutter/flutter/issues/70124 if you have parents also deciding whether to cache a layer then you can end up in situations where a picture layer wants to cache its picture, but it is inside a parent layer that wants to cache its layer. At that point you end up with double-caching.

This was first noted when we added caching to Opacity and ImageFiltered layers, but we have not since introduced a solution for it. If we start caching the TransformLayers  then we might see this situation become more common. Unfortunately, I'm not sure we have a good measure of this happening, but you might spy it in the timeline traces if you see a TL see a nested child fill the cache right before the TL itself fills the cache.

I need to create an Issue for redesigning our cache strategy. We really should not be caching during PreRoll. The reason we do that is to avoid render target switching in the Paint phase, but Preroll is too early to look at the larger picture and plan the cache entries. We really need Preroll to simply identify the nodes that are good targets for caching, then another pass that chooses which to cache using a global strategy once the candidates have been identified, then a final pass to Paint. Another problem this would solve is that evictions happen after we fill the new entries because the new entries are made before we are done with the Preroll scan. This means we have to create new cache entries before we evict expired ones. If this was a separate pass then the evictions could happen before the global decision to add new entries.

My recommendation would be to investigate if the RepaintBoundary objects are too low in the list to provide the best response, raise them higher so that the scrollable container contains only RepaintBoundaries as children and compare to these results.

Then we really need to get started on redesigning the caching process and take it out of Preroll...

**Comment by dnfield** [comment]
I think if we just reverted https://github.com/flutter/flutter/pull/29701/ it would work. Not sure what impact that has on goldens. We could probably go even further and take out the possibility ever pushing a layer.

Part of why this was done too was to help with debugging fuchsia-specific needs on non-fuchsia platforms, but the code for that has since been removed.

**Comment by flar** [comment]
> > The scrollable container should literally be a list of RepaintBoundaries
> 
> The scrollable container is a list of repaint boundaries already.

I eventually saw that. It was the PhysicalShapeLayer that gets pushed into the RepaintBoundary that is disrupting caching the complete scrollable item.

> Even if we move PhysicalShape back to not pushing a layer, one can imagine other situations where it may be beneficial to cache larger chunks, e.g. SVGs will soon be their own picture and it may still make sense to cache them as a larger unit with other stuff around them.

That could be handled by https://github.com/flutter/engine/pull/29155

> > There is no evaluation of complexity or stability, the cache entry is populated immediately on the first try
> 
> In my current implementation I only start caching when the content of the offset layer has been stable for three frames, similar to ImageFilterLayer.

I haven't seen the code for this experiment, but in looking back at image_filter_layer I can see that it does its own stability test.

**Comment by flar** [comment]
One of the tasks I've identified for improving our RasterCache is to have a separate pass to evaluate all of the candidates for caching. With such a mechanism we would likely not need to specifically tag individual layers as candidates and the list could be inferred from examining the stability of sub-trees. This may not happen soon (it would benefit a lot from the as-yet-disabled DiffContext code which will be enabled soon).

See https://github.com/flutter/flutter/issues/91923

**Comment by goderbauer** [comment]
I've some additional data from an internal use case at b/203469933, which looks less promising.

**Comment by ColdPaleLight** [comment]
I think that TransformLayer participating in raster caching has significantly improved the fluency of list scrolling scenarios.

But I prefer to let the user specify whether RepaintBoundary needs to participate in raster caching, as described in this Issue https://github.com/flutter/flutter/issues/32662.

I wrote experimental PRs to verify whether the performance improved, and from the benchmark results, it did reduce raster times.

PR:
engine:  https://github.com/flutter/engine/pull/32615
framework: https://github.com/flutter/flutter/pull/101754

Test: complex_layout - scroll_perf  -  complex_layout_scroll_perf.timeline_summary
Device: Nexus 6P, Android 8.1.0

Run 1 with cached RepaintBoundary:
```
 "average_frame_build_time_millis": 2.22721359223301,
  "90th_percentile_frame_build_time_millis": 3.684,
  "99th_percentile_frame_build_time_millis": 14.971,
  "worst_frame_build_time_millis": 22.498,
  "missed_frame_build_budget_count": 1,
  "average_frame_rasterizer_time_millis": 7.6445673076923075,
  "90th_percentile_frame_rasterizer_time_millis": 10.271,
  "99th_percentile_frame_rasterizer_time_millis": 16.875,
  "worst_frame_rasterizer_time_millis": 18.243,
  "average_vsync_transitions_missed": 1.0,
  "90th_percentile_vsync_transitions_missed": 1.0,
  "99th_percentile_vsync_transitions_missed": 1.0,
  "average_vsync_frame_lag": 275.80582524271847,
  "90th_percentile_vsync_frame_lag": 428.0,
  "99th_percentile_vsync_frame_lag": 651.0,
  "average_layer_cache_count": 3.1619047619047618,
  "90th_percentile_layer_cache_count": 4.0,
  "99th_percentile_layer_cache_count": 4.0,
  "worst_layer_cache_count": 4.0,
  "average_layer_cache_memory": 22.813605028571406,
  "90th_percentile_layer_cache_memory": 31.022068,
  "99th_percentile_layer_cache_memory": 31.027309,
  "worst_layer_cache_memory": 31.027309,
  "average_picture_cache_count": 1.019047619047619,
  "90th_percentile_picture_cache_count": 1.0,
  "99th_percentile_picture_cache_count": 2.0,
  "worst_picture_cache_count": 2.0,
  "average_picture_cache_memory": 1.4139638476190455,
  "90th_percentile_picture_cache_memory": 1.373291,
  "99th_percentile_picture_cache_memory": 3.280735,
  "worst_picture_cache_memory": 3.736496,
  "total_ui_gc_time": 18.855,
  "30hz_frame_percentage": 0.0,
  "60hz_frame_percentage": 100.0,
  "80hz_frame_percentage": 0.0,
  "90hz_frame_percentage": 0.0,
  "120hz_frame_percentage": 0.0,
  "illegal_refresh_rate_frame_count": 0
```

Run 2 with cached RepaintBoundary:
```
 "average_frame_build_time_millis": 2.083490384615385,
  "90th_percentile_frame_build_time_millis": 4.122,
  "99th_percentile_frame_build_time_millis": 12.742,
  "worst_frame_build_time_millis": 28.832,
  "missed_frame_build_budget_count": 1,
  "average_frame_rasterizer_time_millis": 7.256451923076928,
  "90th_percentile_frame_rasterizer_time_millis": 10.593,
  "99th_percentile_frame_rasterizer_time_millis": 14.371,
  "worst_frame_rasterizer_time_millis": 15.608,
  "missed_frame_rasterizer_budget_count": 0,
  "frame_count": 104,
  "frame_rasterizer_count": 104,
  "new_gen_gc_count": 4,
  "old_gen_gc_count": 2,
  "average_vsync_transitions_missed": 1.0,
  "90th_percentile_vsync_transitions_missed": 1.0,
  "99th_percentile_vsync_transitions_missed": 1.0,
  "average_vsync_frame_lag": 257.59223300970876,
  "90th_percentile_vsync_frame_lag": 376.0,
  "99th_percentile_vsync_frame_lag": 1221.0,
  "average_layer_cache_count": 3.142857142857143,
  "90th_percentile_layer_cache_count": 4.0,
  "99th_percentile_layer_cache_count": 4.0,
  "worst_layer_cache_count": 4.0,
  "average_layer_cache_memory": 22.614155866666646,
  "90th_percentile_layer_cache_memory": 31.022068,
  "99th_percentile_layer_cache_memory": 31.027309,
  "worst_layer_cache_memory": 31.027309,
  "average_picture_cache_count": 1.019047619047619,
  "90th_percentile_picture_cache_count": 1.0,
  "99th_percentile_picture_cache_count": 2.0,
  "worst_picture_cache_count": 2.0,
  "average_picture_cache_memory": 1.4139638476190455,
  "90th_percentile_picture_cache_memory": 1.373291,
  "99th_percentile_picture_cache_memory": 3.280735,
  "worst_picture_cache_memory": 3.736496,
  "total_ui_gc_time": 93.845,
  "30hz_frame_percentage": 0.0,
  "60hz_frame_percentage": 100.0,
  "80hz_frame_percentage": 0.0,
  "90hz_frame_percentage": 0.0,
  "120hz_frame_percentage": 0.0,
  "illegal_refresh_rate_frame_count": 0
```

Run 3 with cached RepaintBoundary:
```
 "average_frame_build_time_millis": 2.4212692307692305,
  "90th_percentile_frame_build_time_millis": 4.11,
  "99th_percentile_frame_build_time_millis": 22.941,
  "worst_frame_build_time_millis": 32.09,
  "missed_frame_build_budget_count": 2,
  "average_frame_rasterizer_time_millis": 7.795737864077668,
  "90th_percentile_frame_rasterizer_time_millis": 10.837,
  "99th_percentile_frame_rasterizer_time_millis": 16.972,
  "worst_frame_rasterizer_time_millis": 17.684,
  "missed_frame_rasterizer_budget_count": 2,
  "frame_count": 104,
  "frame_rasterizer_count": 103,
  "new_gen_gc_count": 6,
  "old_gen_gc_count": 4,
  "average_vsync_transitions_missed": 1.0,
  "90th_percentile_vsync_transitions_missed": 1.0,
  "99th_percentile_vsync_transitions_missed": 1.0,
  "average_vsync_frame_lag": 310.63106796116506,
  "90th_percentile_vsync_frame_lag": 465.0,
  "99th_percentile_vsync_frame_lag": 1385.0,
  "average_layer_cache_count": 3.1634615384615383,
  "90th_percentile_layer_cache_count": 4.0,
  "99th_percentile_layer_cache_count": 4.0,
  "worst_layer_cache_count": 4.0,
  "average_layer_cache_memory": 22.763001192307676,
  "90th_percentile_layer_cache_memory": 31.022068,
  "99th_percentile_layer_cache_memory": 31.027309,
  "worst_layer_cache_memory": 31.027309,
  "average_picture_cache_count": 1.0192307692307692,
  "90th_percentile_picture_cache_count": 1.0,
  "99th_percentile_picture_cache_count": 2.0,
  "worst_picture_cache_count": 2.0,
  "average_picture_cache_memory": 1.4143549326923055,
  "90th_percentile_picture_cache_memory": 1.373291,
  "99th_percentile_picture_cache_memory": 3.280735,
  "worst_picture_cache_memory": 3.736496,
  "total_ui_gc_time": 105.583,
  "30hz_frame_percentage": 0.0,
  "60hz_frame_percentage": 100.0,
  "80hz_frame_percentage": 0.0,
  "90hz_frame_percentage": 0.0,
  "120hz_frame_percentage": 0.0,
  "illegal_refresh_rate_frame_count": 0
```

Run 1 without change:
```
 "average_frame_build_time_millis": 2.5424795918367344,
  "90th_percentile_frame_build_time_millis": 3.826,
  "99th_percentile_frame_build_time_millis": 23.666,
  "worst_frame_build_time_millis": 24.468,
  "missed_frame_build_budget_count": 2,
  "average_frame_rasterizer_time_millis": 9.607404040404035,
  "90th_percentile_frame_rasterizer_time_millis": 12.528,
  "99th_percentile_frame_rasterizer_time_millis": 17.106,
  "worst_frame_rasterizer_time_millis": 19.337,
  "missed_frame_rasterizer_budget_count": 3,
  "frame_count": 98,
  "frame_rasterizer_count": 99,
  "new_gen_gc_count": 6,
  "old_gen_gc_count": 2,
  "average_vsync_transitions_missed": 1.0588235294117647,
  "90th_percentile_vsync_transitions_missed": 1.0,
  "99th_percentile_vsync_transitions_missed": 2.0,
  "average_vsync_frame_lag": 375.83673469387753,
  "90th_percentile_vsync_frame_lag": 468.0,
  "99th_percentile_vsync_frame_lag": 966.0,
  "average_layer_cache_count": 0.0,
  "90th_percentile_layer_cache_count": 0.0,
  "99th_percentile_layer_cache_count": 0.0,
  "worst_layer_cache_count": 0.0,
  "average_layer_cache_memory": 0.0,
  "90th_percentile_layer_cache_memory": 0.0,
  "99th_percentile_layer_cache_memory": 0.0,
  "worst_layer_cache_memory": 0.0,
  "average_picture_cache_count": 2.93,
  "90th_percentile_picture_cache_count": 3.0,
  "99th_percentile_picture_cache_count": 4.0,
  "worst_picture_cache_count": 4.0,
  "average_picture_cache_memory": 5.496746089999995,
  "90th_percentile_picture_cache_memory": 5.64394,
  "99th_percentile_picture_cache_memory": 7.551384,
  "worst_picture_cache_memory": 7.551384,
  "total_ui_gc_time": 22.998,
  "30hz_frame_percentage": 0.0,
  "60hz_frame_percentage": 100.0,
  "80hz_frame_percentage": 0.0,
  "90hz_frame_percentage": 0.0,
  "120hz_frame_percentage": 0.0,
  "illegal_refresh_rate_frame_count": 0
```

Run 2 without change:
```
 "average_frame_build_time_millis": 2.4985252525252526,
  "90th_percentile_frame_build_time_millis": 3.356,
  "99th_percentile_frame_build_time_millis": 24.79,
  "worst_frame_build_time_millis": 26.526,
  "missed_frame_build_budget_count": 2,
  "average_frame_rasterizer_time_millis": 8.989767676767677,
  "90th_percentile_frame_rasterizer_time_millis": 11.335,
  "99th_percentile_frame_rasterizer_time_millis": 16.876,
  "worst_frame_rasterizer_time_millis": 18.44,
  "missed_frame_rasterizer_budget_count": 2,
  "frame_count": 99,
  "frame_rasterizer_count": 99,
  "new_gen_gc_count": 6,
  "old_gen_gc_count": 0,
  "average_vsync_transitions_missed": 1.1111111111111112,
  "90th_percentile_vsync_transitions_missed": 1.0,
  "99th_percentile_vsync_transitions_missed": 2.0,
  "average_vsync_frame_lag": 316.2828282828283,
  "90th_percentile_vsync_frame_lag": 423.0,
  "99th_percentile_vsync_frame_lag": 745.0,
  "average_layer_cache_count": 0.0,
  "90th_percentile_layer_cache_count": 0.0,
  "99th_percentile_layer_cache_count": 0.0,
  "worst_layer_cache_count": 0.0,
  "average_layer_cache_memory": 0.0,
  "90th_percentile_layer_cache_memory": 0.0,
  "99th_percentile_layer_cache_memory": 0.0,
  "worst_layer_cache_memory": 0.0,
  "average_picture_cache_count": 2.93,
  "90th_percentile_picture_cache_count": 3.0,
  "99th_percentile_picture_cache_count": 4.0,
  "worst_picture_cache_count": 4.0,
  "average_picture_cache_memory": 5.496746089999995,
  "90th_percentile_picture_cache_memory": 5.64394,
  "99th_percentile_picture_cache_memory": 7.551384,
  "worst_picture_cache_memory": 7.551384,
  "total_ui_gc_time": 25.985,
  "30hz_frame_percentage": 0.0,
  "60hz_frame_percentage": 100.0,
  "80hz_frame_percentage": 0.0,
  "90hz_frame_percentage": 0.0,
  "120hz_frame_percentage": 0.0,
  "illegal_refresh_rate_frame_count": 0
```

Run 3 without change:
```
 "average_frame_build_time_millis": 2.5092424242424234,
  "90th_percentile_frame_build_time_millis": 4.006,
  "99th_percentile_frame_build_time_millis": 23.334,
  "worst_frame_build_time_millis": 28.724,
  "missed_frame_build_budget_count": 2,
  "average_frame_rasterizer_time_millis": 9.330151515151515,
  "90th_percentile_frame_rasterizer_time_millis": 13.054,
  "99th_percentile_frame_rasterizer_time_millis": 18.297,
  "worst_frame_rasterizer_time_millis": 18.703,
  "missed_frame_rasterizer_budget_count": 4,
  "frame_count": 99,
  "frame_rasterizer_count": 99,
  "new_gen_gc_count": 6,
  "old_gen_gc_count": 2,
  "average_vsync_transitions_missed": 1.0,
  "90th_percentile_vsync_transitions_missed": 1.0,
  "99th_percentile_vsync_transitions_missed": 1.0,
  "average_vsync_frame_lag": 334.07142857142856,
  "90th_percentile_vsync_frame_lag": 459.0,
  "99th_percentile_vsync_frame_lag": 1021.0,
  "average_layer_cache_count": 0.0,
  "90th_percentile_layer_cache_count": 0.0,
  "99th_percentile_layer_cache_count": 0.0,
  "worst_layer_cache_count": 0.0,
  "average_layer_cache_memory": 0.0,
  "90th_percentile_layer_cache_memory": 0.0,
  "99th_percentile_layer_cache_memory": 0.0,
  "worst_layer_cache_memory": 0.0,
  "average_picture_cache_count": 2.91,
  "90th_percentile_picture_cache_count": 3.0,
  "99th_percentile_picture_cache_count": 4.0,
  "worst_picture_cache_count": 4.0,
  "average_picture_cache_memory": 5.4585972099999935,
  "90th_percentile_picture_cache_memory": 5.64394,
  "99th_percentile_picture_cache_memory": 7.551384,
  "worst_picture_cache_memory": 7.551384,
  "total_ui_gc_time": 30.912,
  "30hz_frame_percentage": 0.0,
  "60hz_frame_percentage": 100.0,
  "80hz_frame_percentage": 0.0,
  "90hz_frame_percentage": 0.0,
  "120hz_frame_percentage": 0.0,
  "illegal_refresh_rate_frame_count": 0
```

**Comment by flar** [comment]
Raster caching transform widgets is already done if you add a filterQuality to their widget.

We also have been steering away from adding specific caching hints to the Flutter classes, so the Issue you mention has been falling into disfavor over the past couple of years. @dnfield @zanderso

**Comment by ColdPaleLight** [comment]
> Raster caching transform widgets is already done if you add a filterQuality to their widget.
> 
> We also have been steering away from adding specific caching hints to the Flutter classes, so the Issue you mention has been falling into disfavor over the past couple of years. @dnfield @zanderso

I learned before that Transform uses ImageFilterLayer instead of TransformLayer after adding filterQuality. What I would like to discuss here is whether or how the TransformLayer corresponding to RepaintBoundary should participate in raster caching in scrolling scenes.

In experiments I found that properly caching the RepaintBoundary content to the raster cache can significantly improve scrolling performance, but it also requires more memory. So I hope the developer will have the option to control which RepaintBoundary they wants to cache and which RepaintBoundary is not.

I also tried to cache all RepaintBoundary contents if it didn't change for three frames, like @goderbauer tried before. I found cost more memory and doesn't bring further scrolling performance improvements. 

It's not clear to me why we should avoid adding options to RepaintBoundary. In the process of developing flutter applications, developers have more useful information, why not give them a chance to find the best balance between scrolling performance and memory cost for their applications?

**Comment by zanderso** [comment]
In terms of the design of the API, my hope is that we can get down to one advisory boolean flag on one widget with which the developer can suggest that raster caching may be beneficial. The Engine would be free to ignore this flag, or to decide to do it eagerly or only after a certain number of frames, and everything else about the actual policy. But in terms of the API design, if we have more than one flag, and the flags are scattered around all over the place, and they mean different things, it becomes both too hard for the developer to understand, and too difficult for the Engine to implement faithfully.

In particular, things like this:

> Raster caching transform widgets is already done if you add a filterQuality to their widget.

are an implementation detail that aren't officially part of the API.

So, I'm in favor of work in this area as long as it's moving us toward a simpler, easier-to-understand API. But if we're just adding more flags with no plans to consolidate, I think we need to put more thought into this.

**Comment by ColdPaleLight** [comment]
I like the idea about the simpler, easier-to-understand API.

It's great that Transform can participate in caching after adding filterQuality. But it can be a bit confusing for developers, especially when they find that the Transform's behavior changes a lot between filterQuality is not set and filterQuality is set to FilterQuality.none.

If Transform can also add a flag named `rasterCache` to decide whether to participate in the cache, and let the responsibility of filterQuality return to its literal meaning, I believe it will be easier for developers to understand. And it can also consolidate the API well.

**Comment by flar** [comment]
> In particular, things like this:
> 
> > Raster caching transform widgets is already done if you add a filterQuality to their widget.
> 
> are an implementation detail that aren't officially part of the API.

The filterQuality property is an official part of the API and was added specifically to enable bitmap operations.

https://api.flutter.dev/flutter/widgets/Transform/filterQuality.html
https://api.flutter.dev/flutter/widgets/ScaleTransition/filterQuality.html
https://api.flutter.dev/flutter/widgets/RotationTransition/filterQuality.html

It even documents this behavior.

**Comment by flar** [comment]
> I like the idea about the simpler, easier-to-understand API.
> 
> It's great that Transform widget can participate in caching after adding filterQuality. But it can be a bit confusing for developers, especially when they found that the Transform's behavior changes a lot between filterQuality is not set and filterQuality is set to FilterQuality.none.
> 
> If Transform can also add a flag named `rasterCache` to decide whether to participate in the cache, and let the responsibility of filterQuality return to its literal meaning, I believe it will be easier for developers to understand. And it can also consolidate the API well.

What "original meaning" would filterQuality have *OTHER* than to enable raster caching? The property is completely meaningless unless caching is happening. Adding another flag to enable caching would end up with 2 flags doing the same thing, one of which can actually describe the quality level you want in the bitmap transform and the other basically having no input whatsoever in the quality that we implement.

Check out this design document about the alternatives that were considered. During a meeting where we discussed this document among the Flutter team the concept of a bitmap hint flag floated to the top, and turning that into the filterQuality both gave the developer a flag to enable the bitmap operation as well as a setting to control its quality.

https://docs.google.com/document/d/1Fxdlf7JsA-yAwc_RNuboMG-vhIhchRBZGvDSwg5qR0A/edit#heading=h.pub7jnop54q0

**Comment by flar** [comment]
> But it can be a bit confusing for developers, especially when they found that the Transform's behavior changes a lot between filterQuality is not set and filterQuality is set to FilterQuality.none.

FilterQuality.none is still bitmap caching. It's just using NearestNeighbor.

`Trasnform.filterQuality = null;` turns it off.

**Comment by ColdPaleLight** [comment]
Hi, @flar 
The most common place developers use `filterQuality` is the `Image` widget, which only represents the image sampling quality (original meaning), which does not mean that something should participate in raster caching, right? The `filterQuality` in `Transform` represents two things. The layer corresponding to `Transform` will participate in raster caching (not original meaning), and what is its image sampling quality (original meaning). As a developer, I guess this is actually not easy to understand. So probably adding a `rasterCache ` flag can make it easier for developers to understand what `filterQuality` does. Even if `filterQuality` must not be null when `rasterCache` is true, `filterQuality` must be null or be ignored when `rasterCache` is false.

But this is not the focus of what I'd like to discuss today. I would like to discuss the following things under this Issue:

1. Should `TransformLayer` participate in raster caching?
2. Should we only let the `TransformLayer` created by `RepaintBoundary` participate in raster caching?
3. Should we let the user specify which `RepaintBoundary` participate in raster caching.

Looking forward to your reply, thanks you :-)

**Comment by flar** [comment]
@ColdPaleLight did you read the design document above?

**Comment by ColdPaleLight** [comment]
> @ColdPaleLight did you read the design document above?

I've read it (probably I didn't fully understand it due to language barrier), I think the problem mentioned in the documentation you have solved by transform's filterQuality. Maybe I have some opinions about the API, but never mind it.

What I'd like to discuss is a completely different issue. In a complex list view scrolling scene, a `RepaintBoundary` will be wrapped by default on the outside of each item, and this `RepaintBoundary` corresponds to a `TransformLayer` at the engine layer. If the items in the list are very complex, and may be composed of multiple layers and pictures, and its content will not change, should we cache this `TransformLayer`?

If we should cache it, then how should we tell engine that ·`TransformLayer` is generated by the `RepaintBoundary`? In addition, developers may not want to use this caching strategy to cache all `RepaintBoundary`. Maybe they only want to cache the `RepaintBoundary` in the list, so should we provide a flag to the user?

**Comment by flar** [comment]
> But this is not the focus of what I'd like to discuss today. I would like to discuss the following things under this Issue:
> 
> 1. Should `TransformLayer` participate in raster caching?

Participate in what way? By directly caching its children? That is only useful in a small set of ways that are already dealt with. Please describe a scenario in which it might "participate".

In particular, the Issue that you are having this discussion in was describing a specific way in which a TransformLayer might "participate in caching", but in the end, it wasn't really the right solution. The description talks about caching children when scrolling them. But, how does it know when it is scrolling the children? Do we add another flag?

If it knows it is a scroll container's translate mechanism, then how should it participate? By caching all of its children? That would make sense if all of the children were present and stable, but they aren't. As you scroll them they constantly come and go which means the TransformLayer is constantly having to recache its children. If it encapsulates them in a RepaintBoundary then they will cache if they are stable anyway, but doing so will end up churning the cache every time an entry in the list appears or disappears. The best answer in this case is for each child to be cached depending on how complicated that child is. And we already do that if they don't change. All of that is handled in PictureLayer and DisplayListLayer.

So, it can't know whether to make this optimization, and if it did, it would have difficulty in doing it correctly with the current layer caching mechanisms, and if it doesn't then the right thing should happen if the developers construct their lists correctly (i.e. insert RepaintBoundaries to keep their list items separate) - which is along the lines of what has been suggested above, but the thing is that you don't need a flag for a RepaintBoundary to cache itself, it just needs to remain stable.

And if you think that developers instinctively know where to put RepaintBoundary objects, the entire reason this issue is here is because the RepaintBoundary objects were in the wrong place in the tree to do their job effectively. Fixing that ran into another issue related to "double caching" which is being [worked on](https://github.com/flutter/engine/pull/31892).

Now, if we expand this to consider allowing TransformLayers to cache even if their transform isn't a translate, then we end up having to have a quality heuristic that decides whether scaling a child by 1.05 looks good enough that the developer won't notice or care. We'll get that wrong because "heuristic" generally means "bug" in most cases. If we do it with linear interpolation, would that bridge the gap? What about mipmapping? For any operation that isn't a translation, we basically need to loop the developer in and they need to specify that it is OK to use a bitmap operation for this one case. That is where the filterQuality parameters come in. They both indicate that the developer is OK with doing a bitmap operation despite the quality loss, and which level of algorithm retains enough quality for their tastes and design goals.

> 2. Should we only let the `TransformLayer` created by `RepaintBoundary` participate in raster caching?

It already does. If it is just a translation, then the child will cache or not cache depending on its complexity which is what we want it to do. If TransformLayer tries to do this manually then 60-70% of the time we will end up with the child being cached twice until we fix the above "double caching" issue.

> 3. Should we let developers specify which `RepaintBoundary` participate in raster caching.

They already use RepaintBoundary for another purpose and they don't do that very well as it stands. If we add a flag to the widget to force caching we'll disrupt a reasonably good mechanism already built into it that does automatic caching, but there are bugs in that to be fixed. The above "double caching" issue is one of them. The other is to improve our [complexity metrics](https://github.com/flutter/engine/pull/31176) so we only cache things that will speed us up.

If developers put RepaintBoundary objects in their trees at good locations then the natural picture caching code will end up turning them into bitmaps pretty effectively. Then simply translating them means that they will draw from the cache on every frame. And there you have a translation Layer "participating in caching" without it having to have a line of code added to it.

**Comment by flar** [comment]
Another reason not to cache all of the children of a scrolling list together...

If any of them contains an animating stamp or other content that changes, it disrupts the caching of all of the content. If each child is cached separately then only the items that have dynamic content will have to re-render on every frame.

**Comment by dnfield** [comment]
I think I'm getting deja vu here but - I'm pretty sure the ask here is not to cache a transform layer that wraps a bunch of scrolling children, but to cache individual children (which are each an individual TransformLayer once we get to C++ land).

**Comment by ColdPaleLight** [comment]
@flar here is the benchmark result of making RepaintBoundary Cacheable https://github.com/flutter/flutter/issues/91648#issuecomment-1096763647

I am using the complex_layout example provided by flutter, and you can see that the average raster time is significantly improved. At the same time, I also observed that the double-cache issue does not occur in this scenario. Because TransformLayer generates a cache in the third `Prepare`, and its internal displaylist has no chance to generate a cache.

I think the current raster caching mechanism is not perfect. Adding a flag to RasterCache will not destroy current caching mechanism, but will give developers the opportunity to tune it.

If we can optimize complex_layout like this, then for real-world flutter applications, I am more confident that developers can optimize through it.

I'll provide a screen recording video later,  that will be more intuitive

**Comment by ColdPaleLight** [comment]
without change
https://user-images.githubusercontent.com/31977171/163114839-52379bb7-c420-4147-93ad-ee79f628e5b8.MP4

cache RepaintBoundary in sliver
https://user-images.githubusercontent.com/31977171/163114786-eed60e2c-c8a1-4acf-b7f6-6fc877b787d1.MOV

**Comment by flar** [comment]
Here's an implementation issue that I think is probably more important than playing with manual caching on various layers. I wrote a simple program to demonstrate the issue.

Currently, when you add a RepaintBoundary to an app, it splits the tree down the middle completely shattering it. You go from 1 "display list" (as the framework documentation calls it, but it really could be a combination of Layers and DisplayLists in implementation) to a new tree that one might expect to have 2 or 3 sections (a before section, a repaint boundary'd section, and an after section). In the case of my app, it goes from 1 to 21 sections. This is due to the way that the Flutter tree handles the properties it refers to as "compositing" properties. If anything in the tree is composited then all parent widgets that might otherwise just draw to the Canvas instead all become "Layers" that cause their parents to also need to "composite".

 <details>
  <summary>Example code!</summary>

```
import 'dart:math';

import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'RepaintBoundary Tree Impact Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'RepaintBoundary Tree Impact Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  bool _useRepaintBoundary = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: _NestedWidget(
          level: 10,
          rotateDegrees: 9,
          useRepaintBoundary: _useRepaintBoundary,
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () => setState(() => _useRepaintBoundary = !_useRepaintBoundary),
        tooltip: '${_useRepaintBoundary ? 'Remove' : 'Add'} RepaintBoundary',
        child: Icon(_useRepaintBoundary ? Icons.remove : Icons.add),
      ),
    );
  }
}

class _NestedWidget extends StatelessWidget {
  const _NestedWidget({
    required this.level,
    required this.rotateDegrees,
    required this.useRepaintBoundary,
  });

  final int level;
  final double rotateDegrees;
  final bool useRepaintBoundary;

  @override
  Widget build(BuildContext context) {
    if (level == 0) {
      Widget child = const SizedBox(width: 200, height: 300);
      if (useRepaintBoundary) {
        child = RepaintBoundary(child: child);
      }
      return child;
    } else {
      return Transform(
        transform: Matrix4.rotationZ(rotateDegrees * pi / 180),
        alignment: Alignment.center,
        child: CustomPaint(
          painter: _MyPainter(color: Colors.blue, isHorizontal: true),
          foregroundPainter: _MyPainter(color: Colors.green, isHorizontal: false),
          child: _NestedWidget(
            level: level - 1,
            rotateDegrees: rotateDegrees,
            useRepaintBoundary: useRepaintBoundary,
          ),
        ),
      );
    }
  }
}

class _MyPainter extends CustomPainter {
  _MyPainter({required this.color, required this.isHorizontal});

  final Color color;
  final bool isHorizontal;

  @override
  void paint(Canvas canvas, Size size) {
    Paint p = Paint()
        ..color = color;
    if (isHorizontal) {
      canvas.drawRect(Rect.fromLTWH(0, size.height / 2 - 5, size.width, 10), p);
    } else {
      canvas.drawRect(Rect.fromLTWH(size.width / 2 - 5, 0, 10, size.height), p);
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => true;
}
```
</details>

An observatory view of it before I click the button to add the RepaintBoundary:

<img width="1147" alt="Screen Shot 2022-04-13 at 12 08 08 AM" src="https://user-images.githubusercontent.com/50503328/163120774-139be350-d335-452c-ac32-a9ddd373f7f3.png">

And another observatory view after I click to add the RepaintBoundary. Note that the RepaintBoundary is only added to the leaf widget!

<img width="1147" alt="Screen Shot 2022-04-13 at 12 09 16 AM" src="https://user-images.githubusercontent.com/50503328/163120862-734067f6-b4df-4771-bffb-9fb6b9095fec.png">

**Comment by flar** [comment]
> I think the current raster caching mechanism is not perfect. Adding a flag to RasterCache will not destroy current caching mechanism, but will give developers the opportunity to tune it.

But what I'm saying is that we are discussing throwing good money after bad rather than fixing the real underlying problems such as seen in the example I gave.

**Comment by flar** [comment]
> I am using the complex_layout example provided by flutter, and you can see that the average raster time is significantly improved. At the same time, I also observed that the double-cache issue does not occur in this scenario. Because TransformLayer generates a cache in the third `Prepare`, and its internal displaylist has no chance to generate a cache.

This is a delicate balance achieved as a side effect of copying the same threshold between two different mechanisms. It is fragile and someone might change the values and break this happy coincidence.

When I first observed the issue with a similar case, the double-caching was only for one frame in many cases. But, still it happened. And does the cache churn when a new item is added to the list?

I looked at the PRs and did not see anywhere you were setting the flag...? How is your proposed mechanism working?

**Comment by ColdPaleLight** [comment]
> > I think the current raster caching mechanism is not perfect. Adding a flag to RasterCache will not destroy current caching mechanism, but will give developers the opportunity to tune it.
> 
> But what I'm saying is that we are discussing throwing good money after bad rather than fixing the real underlying problems such as seen in the example I gave.

Regarding RepaintBoundary, I think your example is a very special case, in your example it does seem to have side effects. But in more real flutter scenarios it is very helpful.

For example, ListView will automatically wrap RepaintBoundary for its children by default for better performance. If the ListView does not wrap the RepaintBoundary for the child, you can see that there will be a big drop in performance.

The following is also the scene of `complex_layout`, I set the `addRepaintBoundaries` of `ListView` to `false`.

1. addRepaintBoundaries is false
```
  "average_frame_build_time_millis": 5.314563106796118,
  "90th_percentile_frame_build_time_millis": 8.034,
  "99th_percentile_frame_build_time_millis": 14.562,
  "worst_frame_build_time_millis": 18.934,
  "missed_frame_build_budget_count": 1,
  "average_frame_rasterizer_time_millis": 9.54069902912622,
  "90th_percentile_frame_rasterizer_time_millis": 11.991,
  "99th_percentile_frame_rasterizer_time_millis": 23.355,
  "worst_frame_rasterizer_time_millis": 25.114,
  "missed_frame_rasterizer_budget_count": 3,
  "frame_count": 103,
  "frame_rasterizer_count": 103,
  "new_gen_gc_count": 8,
  "old_gen_gc_count": 2,
  "average_vsync_transitions_missed": 1.0,
  "90th_percentile_vsync_transitions_missed": 1.0,
  "99th_percentile_vsync_transitions_missed": 1.0,
  "average_vsync_frame_lag": 267.7378640776699,
  "90th_percentile_vsync_frame_lag": 393.0,
  "99th_percentile_vsync_frame_lag": 474.0,
  "average_layer_cache_count": 0.0,
  "90th_percentile_layer_cache_count": 0.0,
  "99th_percentile_layer_cache_count": 0.0,
  "worst_layer_cache_count": 0.0,
  "average_layer_cache_memory": 0.0,
  "90th_percentile_layer_cache_memory": 0.0,
  "99th_percentile_layer_cache_memory": 0.0,
  "worst_layer_cache_memory": 0.0,
  "average_picture_cache_count": 1.0,
  "90th_percentile_picture_cache_count": 1.0,
  "99th_percentile_picture_cache_count": 1.0,
  "worst_picture_cache_count": 1.0,
  "average_picture_cache_memory": 1.3732909999999978,
  "90th_percentile_picture_cache_memory": 1.373291,
  "99th_percentile_picture_cache_memory": 1.373291,
  "worst_picture_cache_memory": 1.373291,
  "total_ui_gc_time": 109.917,
  "30hz_frame_percentage": 0.0,
  "60hz_frame_percentage": 100.0,
  "80hz_frame_percentage": 0.0,
  "90hz_frame_percentage": 0.0,
  "120hz_frame_percentage": 0.0,
  "illegal_refresh_rate_frame_count": 0
```

2. addRepaintBoundaries is true (default)

```
"average_frame_build_time_millis": 2.5092424242424234,
  "90th_percentile_frame_build_time_millis": 4.006,
  "99th_percentile_frame_build_time_millis": 23.334,
  "worst_frame_build_time_millis": 28.724,
  "missed_frame_build_budget_count": 2,
  "average_frame_rasterizer_time_millis": 9.330151515151515,
  "90th_percentile_frame_rasterizer_time_millis": 13.054,
  "99th_percentile_frame_rasterizer_time_millis": 18.297,
  "worst_frame_rasterizer_time_millis": 18.703,
  "missed_frame_rasterizer_budget_count": 4,
  "frame_count": 99,
  "frame_rasterizer_count": 99,
  "new_gen_gc_count": 6,
  "old_gen_gc_count": 2,
  "average_vsync_transitions_missed": 1.0,
  "90th_percentile_vsync_transitions_missed": 1.0,
  "99th_percentile_vsync_transitions_missed": 1.0,
  "average_vsync_frame_lag": 334.07142857142856,
  "90th_percentile_vsync_frame_lag": 459.0,
  "99th_percentile_vsync_frame_lag": 1021.0,
  "average_layer_cache_count": 0.0,
  "90th_percentile_layer_cache_count": 0.0,
  "99th_percentile_layer_cache_count": 0.0,
  "worst_layer_cache_count": 0.0,
  "average_layer_cache_memory": 0.0,
  "90th_percentile_layer_cache_memory": 0.0,
  "99th_percentile_layer_cache_memory": 0.0,
  "worst_layer_cache_memory": 0.0,
  "average_picture_cache_count": 2.91,
  "90th_percentile_picture_cache_count": 3.0,
  "99th_percentile_picture_cache_count": 4.0,
  "worst_picture_cache_count": 4.0,
  "average_picture_cache_memory": 5.4585972099999935,
  "90th_percentile_picture_cache_memory": 5.64394,
  "99th_percentile_picture_cache_memory": 7.551384,
  "worst_picture_cache_memory": 7.551384,
  "total_ui_gc_time": 30.912,
  "30hz_frame_percentage": 0.0,
  "60hz_frame_percentage": 100.0,
  "80hz_frame_percentage": 0.0,
  "90hz_frame_percentage": 0.0,
  "120hz_frame_percentage": 0.0,
  "illegal_refresh_rate_frame_count": 0
```

**Comment by ColdPaleLight** [comment]
> > I am using the complex_layout example provided by flutter, and you can see that the average raster time is significantly improved. At the same time, I also observed that the double-cache issue does not occur in this scenario. Because TransformLayer generates a cache in the third `Prepare`, and its internal displaylist has no chance to generate a cache.
> 
> This is a delicate balance achieved as a side effect of copying the same threshold between two different mechanisms. It is fragile and someone might change the values and break this happy coincidence.
> 
> When I first observed the issue with a similar case, the double-caching was only for one frame in many cases. But, still it happened. And does the cache churn when a new item is added to the list?

I believe the double-caching issue will be resolved in the near future, and there are already PRs working towards it.

> I looked at the PRs and did not see anywhere you were setting the flag...? How is your proposed mechanism working?

In the following two lines of code in `Sliver`, I set `rasterCache` to `true` for benchmarking. This was not added to the PR, just used when testing locally.
https://github.com/flutter/flutter/blob/31191dd29ae3e2bb794abdccc8ccfa5093b86c02/packages/flutter/lib/src/widgets/sliver.dart#L480

https://github.com/flutter/flutter/blob/31191dd29ae3e2bb794abdccc8ccfa5093b86c02/packages/flutter/lib/src/widgets/sliver.dart#L744

**Comment by ColdPaleLight** [comment]
@flar Another interesting and reasonable thing is that if developers use `Opacity` to wrap the original `ListView` item and set the `opacity` value to `0.99`, the raster times will be optimized. This is thanks to the caching mechanism of `OpacityLayer`.
```
Opacity(
  opacity: 0.99,
  child: originItem
);
```

Of course this is a not recommended solution. But this also proves from another aspect that developers need a correct solution.

**Comment by flar** [comment]
I got a pretty good performance increase simply from drawing the PhysicalShapeLayers instead of adding them as a layer similar to what was suggested above in [Dan's comment](https://github.com/flutter/flutter/issues/91648#issuecomment-943755423). Basically, I did a quick and dirty revert of https://github.com/flutter/flutter/pull/29701 as suggested. Also, I'm not sure I got all of the PSL's that were being used - I just rewrote the one that I saw in a stack trace, but the PR above modified multiple points in the code.

My point is that there is a bug in Flutter that affects the scrolling performance of this demo by blocking effective caching of the list items. Instead of fixing that, you are adding a new heavyweight solution to work around it when it should already be working just fine.

What you are seeing with your fix is that we are going from very incompetent caching of the children via RepaintBoundary objects and bugs that thwart them, to a very rigid forced cache through your new mechanism. You are verifying that caching will help scrolling, but another solution is to fix these bugs and we'll see true scrolling with cached items working they way it was intended and then we can see how it compares to your scrolling with the whole scrollable cached.

**Comment by ColdPaleLight** [comment]
@flar 
I don't think this issue is equivalent to the bug in PhysicalShape. Even if we fix the bug in PhysicalShape, there is no guarantee that a complex ListView's Item whose content will not change participates in raster caching. see https://github.com/flutter/flutter/issues/91648#issuecomment-1098137343 . 

If we say that we cannot guarantee that a complex and stable Item of ListView can be raster cached, then can we give the developer a chance to tune it through benchmarks to see if the Item of ListView needs to be complete raster cache? (If we solve the double-caching issue)

**Comment by flar** [comment]
> Regarding RepaintBoundary, I think your example is a very special case, in your example it does seem to have side effects. But in more real flutter scenarios it is very helpful.

I'm sorry I haven't responded to this issue recently, I have a lot of issues that I'm mulling over.

First, of course RepaintBoundaries add value, I never said they didn't. Everything you pointed out was an example of how they add the specific value that they were designed to add.

My point is that they also have side effects. My example was extreme not because I think that is a common layout, but because it demonstrates the burden the RPB places on the rest of the system. That widget does provide stability for the children, but the way in which it is implemented also fractures the tree above it. I don't think we need that side effect to achieve the same results. A different solution could provide the results without the fracturing.

I'm pointing out that the mechanisms that we are so focused on aren't perfect.

**Comment by flar** [comment]
> In the following two lines of code in `Sliver`, I set `rasterCache` to `true` for benchmarking. Thi was not added to the PR, just used when testing locally.
> 
> https://github.com/flutter/flutter/blob/31191dd29ae3e2bb794abdccc8ccfa5093b86c02/packages/flutter/lib/src/widgets/sliver.dart#L480
> 
> https://github.com/flutter/flutter/blob/31191dd29ae3e2bb794abdccc8ccfa5093b86c02/packages/flutter/lib/src/widgets/sliver.dart#L744

Rather than force a raster cache, why not look at what is preventing the existing caching and fix that.

TransformLayer (and the ImageFilteredLayer used for cached transforms) should be transparent to caching, but currently it is not. It doesn't need to do its own caching, it just needs to not get in the way of existing caching mechanisms.

I already have a PR, still in a draft stage awaiting some more intensive tests, that will make TransformLayer and ImageFiltered transparent to opacity peephole optimization which would make many of the use cases I've seen start to behave without any further changes. In particular, the Android Q transition which is an OpacityLayer around a Transform/ImageFilterLayer will pass through the opacity peephole optimization.

Opacity around a scrolling list would transfer all of the opacity handling to the children who would apply it transparently to their cached version without Opacity needing a layer.

> I think if possible, I hope that `ListView` can also add a flag to specify whether the `RepaintBoundary` generated in `Sliver` needs to be set to `rasterCache`

I don't see a need for a flag in RepaintBoundary if we just fix the existing mechanisms to work as they were originally intended.

**Comment by flar** [comment]
> However, this mechanism is very fragile. As you mentioned in Issue #101810, the tree can be shattered by `RepaintBoundary`. Actually, `Opacity`, `Transform` and many other widgets can shatter tree, resulting in a large `DisplayList` being split into multiple small `DisplayList`s, which may not be able to participate in RasterCache.

Opacity should check if its children need compositing and only use a layer in that case. The else part of that decision would just do a saveLayer with an alpha Paint.

Transform actually does the test I just mentioned and only pushes a layer if its children need compositing.

This "compositing" flag is the "split the tree up the middle" flag. Each child that sets it causes all of its parents to switch to layers instead of rendering commands.

**Comment by flar** [comment]
> for example. If you add an `Opacity` widget below the following two lines of code, then most of the contents of items will no longer be able to be cached in the raster cache, and the performance will degrade again. The reason for this is that the complex DisplayList is split into multiple DisplayLists that are not complex enough.

That's a bug. I submitted an issue to fix it: https://github.com/flutter/flutter/issues/102179

**Comment by flar** [comment]
> @flar Even if we fix the bug in PhysicalShape, there is no guarantee that a complex ListView's Item whose content will not change participates in raster caching. see [#91648 (comment)](https://github.com/flutter/flutter/issues/91648#issuecomment-1098137343) .

That's a bug. We should be able to handle that case.

> If we say that we cannot guarantee that a complex and stable Item of ListView can be raster cached, then can we give developers a chance to tune it through benchmarks to see if the Item of ListView needs to be complete raster cache? (If we solve the double-caching issue)

We should be able to (or strive to) guarantee tree stability rather than give more rope for developers to tie themselves up and prevent future optimizations on our part.

**Comment by ColdPaleLight** [comment]
@flar 
Thanks for your reply, I've learned a lot from you while discussing this issue.

I have some different opinions on this point of yours
> We should be able to (or strive to) guarantee tree stability rather than give more rope for developers to tie themselves up and prevent future optimizations on our part.

Now suppose we have fixed the Opacity and PhysicalShape bugs. But I believe there are still many situations where we can fall into the original problem. For example, replace `Opacity` in https://github.com/flutter/flutter/issues/91648#issuecomment-1098137343 with `RepaintBoundary`, or a horizontal scrolling `ListView` (by default, every Item inside it will be wrapped by `RepaintBoundary`), or replace it with a `Transform` with a `image_filter`, or `ColorFiltered` and many more. I believe both will make the current scrolling problem of `complex_layout` reappear.

I believe that the original intention of @dnfield  to file this issue is to make the content of `RepaintBoundary` be cached as a larger unit.

It looks like you don't like adding a `rasterCache` flag. So can we do this without adding the flag? For example, If we find that the content inside `TransformLayer` is more complicated, and it has been stabilized for 3 frames, and its matrix is translate-only.

**Comment by ColdPaleLight** [comment]
related issue: https://github.com/flutter/flutter/issues/103162

**Comment by jonahwilliams** [comment]
So its a bit hard to track the complete discussion of this issue, but I think I can try to summarize a few points (and @flar can correct me if I'm wrong).

We're aware that it some cases, we create more layers than necessary. We're in the process of fixing this with PhysicalShapeLayer and Opacity, and in some external packages like visibility_detector.

In the case where a picture is significantly split up, raster caching at the layer level can ameliorate some performance problems. However, I think @flar , myself, and others would rather reduce the amount of compositing (and caching) required to render a fast frame than allow more caching to cover up the bad heursitics in the framework/engine.

To that end, I don't think we'd accept https://github.com/flutter/engine/pull/32615 or the related framework PR. We'd rather not have APIs like this that will more or less tie our hands for future optimizations. While we may have originally considered a solution like that in the past, we've gotten good results from removing forced compositing so we're going to continue on that route for now.

**Comment by ColdPaleLight** [comment]
I have closed the relevant PR, so is this issue still a valid issue? It seems that we would not intend to let `TransformLayer` participate in raster caching. Should we close this issue?

**Comment by jonahwilliams** [comment]
Sounds good, thank you @ColdPaleLight
